### DISCUSSION TITLE
One Package vs Many

### DISCUSSION THREAD

**Discussion by CraigSiemens** [pointfreeco/isowords#115] [Q&A]
I noticed the project has one big package with multiple targets. I was wondering what you think the trade-offs might if using more smaller packages instead of one larger one. This is what I've noticed so far but was curuisous if there were any more I was missing.

### Pros
- Allows keeping the related sources and tests close together in the folder hierarchy and in the Xcode navigator
- Can group together related packages
- May be simpler to pull packages out into their own repos later.

### Cons
- it'd make you duplicate the conditional logic in the package into multiple packages
- swift errors if it gets a package that has no targets

### Unknowns
- Is Xcode able to resolve multiple packages depending on another packages based on the path? I'd assume so but never tried it.

**Comment by mbrandonw** [comment]
Hey @CraigSiemens, for us it was mainly to reduce maintenance burden. One SPM package means one directory structure, one Package.swift, no cognitive overhead to figure out which library belongs to which package, and it's super easy to have modules depend on each other when they are in the same package rather than spread across many packages.

We haven't yet seen a clear reason why multiple SPM packages would be preferred. The pros that you list above are, in my opinion, a little theoretical. Multiple packages _will_ group things together, but that comes with the burden of knowing what exactly you want to group. Depending on how granular or broad you want to be with the packages you may just end up with a lot more churn of moving libraries around to different packages. And pulling packages out into their own repo is a rare event (the majority of our libraries are specific to just isowords), and so we don't think we should optimize for that use case.

These of course are just our opinions. Whether you use multiple packages or a single package, the most important thing is to modularize your code base 😄
